### PR TITLE
Allow for multiple css queries inside eachother

### PR DESCRIPTION
**Pull request by TobiasRoikjer**
## Feature
Chaining css selectors, e.g. `query("webview css:'iframe' css:'button'")

## Tests
https://testcloud.xamarin.com/test/xtc-android-sample_37e26e77-df58-4c8e-9709-c94aa833d1c7/?step=0_0_2&tab=stacktrace

### CODE REVIEW

**Comment by TobiasRoikjer** [comment]
@sapieneptus Alright, this is temporary. The javascript implementation is my idea for how to do this gracefully

**Comment by sapieneptus** [comment]
So is the idea here that you can do arbitrary nested queries of any css result, including iframes?

e.g.
```ruby
query "WebView css:'body' css:'p'
```
?

It would also be great for myself (and probably @jmoody as well) if you could describe what the control flow is generally, and what changes would need to be made on the server / client to accommodate the js changes.

**Comment by TobiasRoikjer** [comment]
Exactly. The most important part is that it is backwards compatible. I think we should discuss this in "person" :)

File: server/calabash-js/src/calabash.js [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
     function isHostMethod (object, property)
     {
-      var t = typeof object[property];
-      return t==='function' ||
-             (!!(t==='object' && object[property])) ||
-             t==='unknown';
+        var t = typeof object[property];
+        return t==='function' ||
+            (!!(t==='object' && object[property])) ||
+            t==='unknown';
```
Comment (jmoody) [inline server/calabash-js/src/calabash.js]: Is this a purely whitespace change?  I know it is a pain, but I would have preferred:

1. Commit the whitespace changes on the main branch and push
2. Rebase the your local branch (before pushing)

**Comment by jmoody** [comment]
Isn't this calabash.js file shared with iOS?  Is this the right place for these changes?

**Comment by sapieneptus** [comment]
The android repo contains copies of the calabash.js files, not the a git submodule.

**Comment by TobiasRoikjer** [comment]
@jmoody I just pushed this to have the javascript public, I haven't really made any effort into gitting this right.

**Comment by jmoody** [comment]
@TobiasRoikjer 

Does this change solve the iframe and x-domain frames problem?

I don't like the idea of nested css selectors, especially if we can solve the iframe / x-domain problems in some other way.

**Comment by sapieneptus** [comment]
@jmoody 

The query language currently won't choke if you have multiple `css` selectors, e.g.
```ruby
query "webView css:'body' css:'h1'"
```
The above simply doesn't work currently. 

My approach allows for a single nested query, provided the first `css` selector is returning an iframe

```ruby
query "webView css:'iframe' css:'h1'"
```

This is partially at @Devneval's suggestion, as it seems the concept of 'switching frame' should already be familiar to an automation engineer. Alternatively, we could alias this special case with a new dedicated syntax:

```ruby
query "webView iframe:'#myIframe' css:'h1'"
```

I don't, however, see a strong case for supporting statement such as
```ruby
query "webView css:'body' css:'div' css:'h1'"
```
since, given a `document`, you should be able to specify any element with a single `css` selector.

**Comment by TobiasRoikjer** [comment]
@sapieneptus this PR would also support css:'body' css:'p' css:'...'

**Comment by Devneval** [comment]
@sapieneptus i would imagine the alias for iframe would be the simplest to understand from an automation engineers point of view as this is most akin to web driver.

```
query "webView iframe:'#myIframe' css:'h1'"
```

**Comment by jmoody** [comment]
+1 for :iframe.


> this PR would also support css:'body' css:'p' css:'...'

I think we don't want to support this.

May I suggest that we close this PR and move this discussion to https://github.com/calabash/calabash-ios-server/pull/308?

**Comment by sapieneptus** [comment]
@TobiasRoikjer yes, that's exactly my point. This PR supports that kind of nesting, which for me is a bit of a sticking point. The approach you've outlined seems like it will achieve that level of flexibility, but I'm afraid we might be expanding the feature too much. 

@jmoody If that's the end of this discussion, then we can continue the review there. However, if possible, I'd like to hear @krukow 's opinion as well.

**Comment by sapieneptus** [comment]
@TobiasRoikjer Did you just merge this with all those debug log statements still in the javascript?

**Comment by TobiasRoikjer** [comment]
@sapieneptus Yea. They don't actually do anything

**Comment by sapieneptus** [comment]
@TobiasRoikjer did the XTC Tests include cross domain and same domain iframes as well? Are there multiple types of webviews we might want to test or is the default webview enough?

File: server/calabash-js/src/calabash.js [hunk 2]
```diff
@@ -134,30 +151,92 @@
...
         }
+
+        return {object: res, window: containerWindow, parent: parentContainer};
     }
-    catch (e)
-    {
-       return JSON.stringify({error:'Exception while running query: '+exp, details:e.toString()})
+
+    function log(message) {
+        //document.getElementById('input').innerHTML += message + "\n";
```
Comment (sapieneptus) [inline server/calabash-js/src/calabash.js]: ?

File: server/calabash-js/src/calabash.js [hunk 3]
```diff
@@ -49,60 +59,67 @@
...
+            res['calSavedIndex'] = addSavedResults(object);
+            log("New res: " + res);
+        } else {
+            log("res is an array");
+        }
+
         return res;
     }
 
+    /* TODO: Fix this, it is broken as `object` is now {object: object, window: x} */
```
Comment (sapieneptus) [inline server/calabash-js/src/calabash.js]: What are the implications of this? Can you explain more?

**Pull request merged**